Add arnold CLI

jmaupetit opened the pull request:
## Purpose

If we want to easily package, distribute and deploy Arnold along with its
applications (aka trays), we need a command line tool dedicated to the task.

## Proposal

- [x] remove legacy scripts
- [x] improve Docker image for production
- [x] add `arnold` CLI
- [x] use the CLI in the CI
- [x] ~update documentation (may be updated in a new PR)~ postponed to #582

jmaupetit committed:
🔥(project) remove legacy scripts

Now that we have a CLI (arnold), it is supposed to be the single
entrypoint to play with arnold.

jmaupetit committed:
✨(project) add arnold CLI

Now that Arnold's deployable applications (aka trays) will be
distributed in separated repositories, we needed a real CLI to play with
them. This CLI (arnold) can be used to bootstrap a new Arnold's project
and/or develop/test trays.

jmaupetit committed:
🔥(project) use arnold -d run to execute dev commands

Instead of having x scripts to run development commands, use a unique
entrypoint: bin/arnold using the -d option to activate the `DEV_MODE`.

An active `DEV_MODE` implies the following behavior:

- `docker run` commands mount the local directory sources as a volume in
  `/app`, or,
- `docker run` commands mount the local application tree in
  `/app/apps/myapp` if a tray as been found in the repository.

Makefile rules have been added to run tests and linters.

jmaupetit committed:
⚡️(project) switch to production vs dev build paradigm

Current work strongly improves docker image best practices that were not
implemented until then:

1. use multi-stages build to separate development and production target
   images,
2. only install development dependencies in the development image,
3. use a slim image to better optimize build size (315MB vs 1.33GB).

jmaupetit committed:
🙈(project) ignore extra local customers

While developing this project, you may want to add a local customer only
used for testing.

jmaupetit committed:
👷(circle) use arnold's CLI in the CI

Instead of using a dedicated bash script in the CI to run playbooks, use
our CLI!

jmaupetit committed:
🔥(project) remove minishift support

Using minishift is no longer the recommended way to start a local
openshift cluster. We now use `oc cluster` to develop locally.

jmaupetit committed:
🚨(playbooks) use capitalized letters for booleans

Fixed missing E604 rule.

jmaupetit committed:
⚡️(project) automate local k8s cluster creation

Use the Makefile to create or delete a local development cluster instead
of relying on another bash script. And use this Make target in the CI.

As env.d environment files are no longer required, I decided to remove
them.

jmaupetit committed:
👷(project) move ci scripts to a single entrypoint

Instead of having multiple CI-related scripts, let's move to a single
entrypoint bash script with options and sub-commands.

jmaupetit commented on bin/arnold:
```diff
@@ -0,0 +1,818 @@
+
+# Docker
+declare -r ARNOLD_IMAGE_NAME="${ARNOLD_IMAGE_NAME:-fundocker/arnold}"
+declare -r ARNOLD_IMAGE_TAG="${ARNOLD_IMAGE_TAG:-5.21.0}"
```
Should be up-to-date.

madmatah commented on bin/activate:
```diff
@@ -0,0 +1,16 @@
+export ANSIBLE_VAULT_PASSWORD="arnold"
+export K8S_AUTH_VERIFY_SSL="no"
+
+echo -e "Development environment activated ✅\\n"
```
Maybe this script want to check that it has been "sourced" ?
Because when executed directly with `bin/activate`, it has no effect.

jmaupetit replied:
I've added a warning message and an error code while trying to execute this script.

madmatah commented on bin/activate:
```diff
@@ -0,0 +1,16 @@
+ARNOLD_IMAGE_TAG="$(tr -d '\n' < VERSION)"
+export ARNOLD_IMAGE_TAG
+export ARNOLD_IMAGE="${ARNOLD_IMAGE_NAME}:${ARNOLD_IMAGE_TAG}"
+export ANSIBLE_VAULT_PASSWORD="arnold"
```
The value of the default password (`"arnold"`) is duplicated 3 times in the project : 
 - bin/activate
 - Makefile
 - CircleCI config

Should it be defined once in a file?

jmaupetit replied:
I'm not sure this is good idea as people might use it without knowing they are using a default vault password. In fact those occurrences in 3 different files serve 3 different (undocumented) purposes:

- the `bin/activate` script is used to activate a development environment to use with the `bin/arnold` script,
- the `Makefile` is only used for development purpose with no other dependencies (see next comment), and,
- the `.circleci/config.yml` file is used for CI usage only.

That said, I'm open to suggestions, what were you thinking of by "defined once in a file"?

madmatah replied:
I was thinking something like this:

------------

1) Add a `arnold_defaults` file containing : 
```
export ARNOLD_DEFAULT_VAULT_PASSWORD="arnold"
```

2) In the Makefile

```
include arnold_defaults

ANSIBLE_VAULT_PASSWORD=$(ARNOLD_DEFAULT_VAULT_PASSWORD)
```

3) In `bin/activate` : 

```
source arnold_defaults
export ANSIBLE_VAULT_PASSWORD="${ARNOLD_DEFAULT_VAULT_PASSWORD}"
```

4) In .circleci/config.yml : 

```
cat arnold_defaults >> $BASH_ENV
echo 'export ANSIBLE_VAULT_PASSWORD="${ARNOLD_DEFAULT_VAULT_PASSWORD}"' >> $BASH_ENV
```

------------

It avoids duplicating the value in the project. And about that :

> I'm not sure this is good idea as people might use it without knowing they are using a default vault password.

I don't think this suggestion makes the problem worse.
If it's a problem, you should print a message saying «Warning, you are using the default arnold password, your vault is not safe».

jmaupetit replied:
Great suggestion! I'll do this. Thanks.

madmatah commented on Makefile:
```diff
@@ -0,0 +1,116 @@
+ARNOLD_IMAGE_TAG_DEV   = $(ARNOLD_IMAGE_TAG)-dev
+ARNOLD_IMAGE           = $(ARNOLD_IMAGE_NAME):$(ARNOLD_IMAGE_TAG)
+ARNOLD_IMAGE_DEV       = $(ARNOLD_IMAGE_NAME):$(ARNOLD_IMAGE_TAG_DEV)
+ANSIBLE_VAULT_PASSWORD = arnold
```
Should it use environment variables instead of these values, when defined ?

```suggestion
ARNOLD_IMAGE_NAME      ?= arnold
ARNOLD_IMAGE_TAG       ?= $(shell tr -d '\n' < VERSION)
ARNOLD_IMAGE_TAG_DEV   ?= $(ARNOLD_IMAGE_TAG)-dev
ARNOLD_IMAGE           ?= $(ARNOLD_IMAGE_NAME):$(ARNOLD_IMAGE_TAG)
ARNOLD_IMAGE_DEV       ?= $(ARNOLD_IMAGE_NAME):$(ARNOLD_IMAGE_TAG_DEV)
ANSIBLE_VAULT_PASSWORD ?= arnold
```

jmaupetit replied:
This is what I did in a first attempt, and, I changed my mind to enforce default development values. Do you have a particular scenario in mind where it could be useful to redefine those thanks to the environment?

madmatah replied:
One thing that comes to my mind is that CircleCI is using `make` in its jobs. And also, it defines these some of these environment variables in `ci_env` task. At this time, the values of these environment variables are the same than the default values defined in the Makefile.

But if someone wants to change the variables defined in `ci_env` job in the future, I'm pretty sure that this will lead to unexpected behavior and confusion.

jmaupetit replied:
Ok then.

madmatah commented on Makefile:
```diff
@@ -0,0 +1,116 @@
+DOCKER_USER = $(DOCKER_UID):$(DOCKER_GID)
+
+# -- k8s
+K8S_DOMAIN="$(shell hostname -I | awk '{print $$1}')"
```
```suggestion
K8S_DOMAIN?="$(shell hostname -I | awk '{print $$1}')"
```

jmaupetit replied:
Ok, maybe this one would be useful :sweat_smile:

jmaupetit commented on bin/arnold:
```diff
@@ -0,0 +1,832 @@
+
+# Flags
+declare -i DEV_MODE=0
+declare -i USE_GPG=0
```
We must be able to bypass this default thanks to the environment:

```suggestion
declare -i USE_GPG=${ARNOLD_USE_GPG:-0}
```

madmatah commented on bin/arnold:
```diff
@@ -0,0 +1,832 @@
+    K8S_AUTH_HOST=$(_get_k8s_auth_host)
+    export K8S_AUTH_HOST
+
+    # Check that the current user is already logged in k8s
```
```suggestion
    # Check that the current user is already logged in k8s. Note that the `whoami`
    # command is only compatible with openshift and should be adapted for k8s.
```

madmatah commented on bin/arnold:
```diff
@@ -0,0 +1,832 @@
+
+    # shellcheck disable=SC2086
+    docker run --rm -ti \
+        -u "$(id -u)" \
```
What do you think about adding a variable like this : 
```
DOCKER_USER="${DOCKER_USER:-$(id -u)}"
```

and use it as a value for the `-u` argument  ?

jmaupetit replied:
In fact, this is breaking the CI as the DOCKER_USER already exists (as a secret)... :sweat_smile:

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,829 @@
+}
+
+function login() {
+    k8s_auth_host=$(_get_k8s_auth_host)
```
You can't have the host if you are not logged in.

This is the output I have when I'm not connected :

```
$ oc version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO
```

jmaupetit replied:
Well spotted! I've changed to the following:

```bash
function _get_k8s_auth_host() {
    echo "https://${K8S_DOMAIN}:8443"
}
```

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,829 @@
+
+    # Get K8S credentials
+    export K8S_DOMAIN
+    export K8S_SA_TOKEN
```
I don't understand these 2 export. Where are they set ? What is exported ? Do we really need to export them ?

jmaupetit replied:
Those are not required. I've cleaned this. :pray:

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,829 @@
+    ${KUBECTL} version | grep Server | awk '{print $2}'
+}
+
+function login() {
```
This function is not listed in the usage function.

jmaupetit replied:
It is not supposed to get called as requiring a service account token, I made it private.

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,829 @@
+    fi
+
+    # Ansible's k8s raw module requires the following two environment
+    # variables to be defined
```
There is only one defined

jmaupetit replied:
Both `K8S_AUTH_API_KEY` and `K8S_AUTH_HOST` environment variables should be defined. I've slightly update my previous change.

lunika commented on .gitignore:
```diff
@@ -16,6 +16,7 @@ env.d/*
 # Tests temporary files
 .coverage
 .pytest_cache
+.pylint.d
```
You should complete your description explaining why you use `-d` and what is `-d` option.

jmaupetit replied:
It should be mentioned in the documentation (README) or do you have something else in mind?

lunika replied:
I mean commit description

jmaupetit replied:
I propose the following:

```
(project) use arnold -d run to execute dev commands

Instead of having x scripts to run development commands, use a unique
entrypoint: bin/arnold using the -d option to activate the `DEV_MODE`.

An active `DEV_MODE` implies the following behavior:

- `docker run` commands mount the local directory sources as a volume in `/app`, or,
- `docker run` commands mount the local application tree in `/app/apps/myapp` if a tray as been found in the repository.

Makefile rules have been added to run tests and linters.
```

WDYT?

lunika replied:
ok for me

lunika commented on .circleci/config.yml:
```diff
@@ -83,23 +80,17 @@ version: 2
-      - *ci_env
       - *docker_login
       - run:
           name: Build arnold production image
```
The name is not accurate anymore with what you are doing

jmaupetit replied:
Do you suggest making it plural? What about:

> Build arnold production & development images

lunika replied:
This is more accurate

lunika reviewed (commented):
![](https://media.giphy.com/media/4GRRBtKrdiFDa/giphy.gif)

jmaupetit commented:
@lunika & @madmatah I've addressed all your comments (thanks for the review) and pushed fixup commits to ease the second round. Ready?

lunika commented on bin/arnold:
```diff
@@ -37,6 +47,8 @@ declare -r CUSTOMER_DIR="${ARNOLD_PROJECT_PATH}/${RELATIVE_CUSTOMER_DIR}"
 declare -r VAULT_PASSWORD_FILE_NAME="password.gpg"
 
 # Binaries
+declare -r CURL="docker run --rm curlimages/curl"
```
No please, don't use docker to run `curl`. Use directly `curl` or `wget`, `wget` is installed everywhere.

jmaupetit replied:
> wget is installed everywhere.

That's a strong statement. I'm not sure to share this thought.

What about trying to use locally installed tools and fall back to the docker image?

jmaupetit replied:
Ok then. After an IRL discussion, `curl` is now a project dependency. Arnold will fail with an explicit error message if it's not installed.

jmaupetit committed:
🎨(project) avoid repeating defaults

The default Arnold vault password used in development and the CI was
repeated multiple times in different configurations. We now use a
centralized defaults configuration to mitigate this issue.

jmaupetit commented:
@lunika & @madmatah I've addressed all your comments. Ready for a new review. :pray:

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+declare -r VAULT_PASSWORD_FILE_NAME="password.gpg"
+
+# Dependencies
+declare -r CURL="curl"
```
I don't see the interest of this variable.

jmaupetit replied:
This is an abstraction layer in case we decide to change our mind and bundle it with Arnold for example. I find it convenient to define all dependencies in the same code block. But of course, this can be changed.

WDYT @madmatah?

lunika replied:
Ok but at the end of the script you added a link to the curl website to install it. What will you do if you change the value of this variable ?

madmatah replied:
I think it's a good practice and consistent with other tools used in this script.

BUT I have a problem with the way it's used line 971 to 975, that limit the possible values here.

jmaupetit replied:
> BUT I have a problem with the way it's used line 971 to 975, that limit the possible values here.

True. Should I remove this check then? What do you suggest?

lunika replied:
removing the `CURL` variable

madmatah replied:
Same here, I would remove the variable and the `eval` L263.

jmaupetit replied:
Done both. Thanks!

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+declare LATEST_VERSION
+
+# Script infos
+declare SCRIPT_NAME
```
Why not declaring and assigning at the same time, like you do after in the script ?

jmaupetit replied:
Shellcheck recommends to do so. See https://github.com/koalaman/shellcheck/wiki/SC2155

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+# Script infos
+declare SCRIPT_NAME
+SCRIPT_NAME=$(basename "$0")
+declare SCRIPT_PATH
```
devito

jmaupetit replied:
Dany.

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+  ["success"]=${COLOR_GREEN}
+  ["info"]=${COLOR_BLUE}
+  ["debug"]=${COLOR_DEFAULT}
+  ["notset"]=${COLOR_DEFAULT}
```
`notset` ? Kesako ?

jmaupetit replied:
I tried to mimic python logging module logic here. See: https://docs.python.org/3/howto/logging.html#logging-levels

lunika replied:
First time I see this... I'm not convinced it's a good thing but why not.

lunika commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+  image      : ${COLOR_BLUE}${ARNOLD_IMAGE}${COLOR_RESET}
+"
+
+LATEST_RELEASE=$(_get_latest_release)
```
What happen if github is down ? Does the script will be usable or not because of this failure ?

jmaupetit replied:
Good point! I should fix that.

jmaupetit replied:
Done

madmatah commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+    echo "${LATEST_RELEASE}" | \
+    ${JQ} .[0].tarball_url
+  )
+  upgraded_bin="/tmp/arnold-${LATEST_VERSION}"
```
I suggest to use `mktemp` to follow best practices in temporary files creation (using $TMPDIR or fallback to /tmp).

```suggestion
  upgraded_bin=$(mktemp -q "arnold-${LATEST_VERSION}-XXXX")
```

Or simply : 

```suggestion
  upgraded_bin=$(mktemp -q)
```

jmaupetit replied:
Of course!

![](https://media.giphy.com/media/1hAxQTH0HEWS3L0oRF/giphy.gif)

madmatah commented on bin/arnold:
```diff
@@ -0,0 +1,1009 @@
+  upgraded_bin=$(_download_latest_release)
+
+  chmod +x "${upgraded_bin}"
+  cp -i "${upgraded_bin}" "${SCRIPT_PATH}"
```
1/ Why `cp` and not `mv` ? No need to clean-up, then ;-)

2/ The `-i` arguments prevent from running an upgrade without a TTY. Is this the desired behaviour?
Maybe add a `--force` option to your upgrade command without interactive confirmation ?

jmaupetit replied:
> Why cp and not mv ? No need to clean-up, then ;-)

As it was a temporary file, I though it wasn't worth a cleanup... but ok then.

> The -i arguments prevent from running an upgrade without a TTY. Is this the desired behaviour?

Yes a kind of lazy confirmation step.

> Maybe add a --force option to your upgrade command without interactive confirmation ?

Good idea.

jmaupetit commented:
Aaaaand fixed. Ping @lunika && @madmatah :pray: 

![](https://media.giphy.com/media/H6PyizSD08ygryW8N6/giphy.gif)

madmatah commented on bin/arnold:
```diff
@@ -0,0 +1,1022 @@
+
+# _ensure_k8s_project_is_set: ensure that k8s's target project has
+# been set via the ARNOLD_CUSTOMER & ARNOLD_ENVIRONMENT environment variables
+function _ensure_k8s_project_is_set() {
```
Minor suggestion to use k8s vocabulary : 

```suggestion
function _ensure_k8s_namespace_is_set() {
```

madmatah reviewed (approved):
Minor suggestions, but the most important thing to note is :

![](https://media.giphy.com/media/Z6f7vzq3iP6Mw/giphy.gif)
